### PR TITLE
 Unpin Xarray: dataArray -> dataArray.variable

### LINKED ISSUE #381: Remove Xarray Private API and remove pinning
**Describe the bug**
Xarray 2023.03.0 breaks tests due to a call on [line 1818 of xarray/core/common.py](https://github.com/pydata/xarray/blob/1e361ccb9123fe25acfd9e3364c911c1eec7d9db/xarray/core/common.py#L1818)

**To Reproduce**
run tests

**Expected behavior**
no error

**Screenshots**
[Failing test run](https://github.com/NCAR/geocat-comp/actions/runs/4503455909/jobs/7926663488)
**OS:**
Both MacOS and Ubuntu

**Environment**
n/a

### PR DESCRIPTION
**Pull request by cyschneck**
## PR Summary

Closes #381

## PR Checklist

**General**
- [x] Make an issue if one doesn't already exist
- [x] Link the issue this PR resolves by adding `closes #XXX` to the PR description where XXX is the number of the issue.
- [x] Add a brief summary of changes to `docs/release-notes.rst` in a relevant section for the next unreleased release. Possible sections include: Documentation, New Features, Bug Fixes, Internal Changes, Breaking Changes/Deprecated
- [x] Add appropriate labels to this PR
- [x] Make your changes in a forked repository rather than directly in this repo
- [x] Open this PR as a draft if it is not ready for review
- [x] Convert this PR from a draft to a full PR before requesting reviewers
- [x] Passes `precommit`. To set up on your local, run `pre-commit install` from the top level of the repository. To manually run pre-commits, use `pre-commit run --all-files` and re-add any changed files before committing again and pushing.
- [x] If needed, squash and merge PR commits into a single commit to clean up commit history

### CODE REVIEW

**Comment by cyschneck** [comment]
Test errors appear to arise in most recent xarray (2023.10.1) interaction with existing tests

`/xarray/core/dataarray.py:869: in __setitem__`
```
self.variable[key] = value
```
Calls: `/xarray/core/variable.py:880: in __setitem__`
```
value = value.set_dims(dims).data
```
Throws: `ValueError: new dimensions ('sigma',) must be a superset of existing dimensions ('dim_0',)`

**Comment by cyschneck** [comment]
Current test failure: [`test/test_climatologies.py::Test_Month_to_Season::test_month_to_season_custom_time_coordinate[ds4-dataset1-None-Tair-expected1:350`](https://github.com/NCAR/geocat-comp/blob/cf476447a8b3893bade867e0d3e69ad80200e557/test/test_climatologies.py#L350C8-L350C8)

Throws the error `ValueError: unit abbreviation w/o a number`, but the actual issue is `loffset` in [climatologies.py:303](https://github.com/NCAR/geocat-comp/blob/cf476447a8b3893bade867e0d3e69ad80200e557/geocat/comp/climatologies.py#L303C5-L305C49) when xarray resample is called

```
means = data_filter.resample({time_coord_name: quarter}, loffset='MS').mean(keep_attrs=keep_attrs)
```

ValueError originates in xarray [`resample_cftime.py:155`](https://github.com/pydata/xarray/blob/fcdc81028bd3c066fc7c9fbf965d2c66fcf8da4c/xarray/core/resample_cftime.py#L155) where xarray calls pandas to_timedelta with `loffset`

```
labels = labels + pd.to_timedelta(self.loffset)
```

`loffset` is accepted since `CFTimeGrouper:first_items` expected `loffset` to be a a string, datetime.timedelta, or instance of a BaseCFTimeOffset. As a string, `loffset="MS" with pandas function ``pd.to_timedelta("MS")`. However, [pandas has dropped `loffset` string support since pandas version 1.1.0](https://pandas.pydata.org/pandas-docs/version/1.1/reference/api/pandas.DataFrame.resample.html). The ValueError being thrown is a deceptive error due to when pandas attempts to convert the loffset "MS" to a timedelta, but as loffset is deprecated within pandas when "MS" is attempted to be [converted to TimeDelta without units](https://github.com/pandas-dev/pandas/blob/eedf0d5cc2801cc3849ec9abb4e8b3f509d91ee9/pandas/core/tools/timedeltas.py#L222)

```
result = Timedelta(r="MS", unit=None)
```

It throws an invalid ValueError. As an invalid Timedelta string, the error can be replicated as `pandas.to_timedelta("MS")`

```
>>> import pandas as pd
>>> pd.to_timedelta("MS")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/miniconda3/envs/geocat_comp_build/lib/python3.11/site-packages/pandas/core/tools/timedeltas.py", line 223, in to_timedelta
    return _coerce_scalar_to_timedelta_type(arg, unit=unit, errors=errors)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/miniconda3/envs/geocat_comp_build/lib/python3.11/site-packages/pandas/core/tools/timedeltas.py", line 233, in _coerce_scalar_to_timedelta_type
    result = Timedelta(r, unit)
             ^^^^^^^^^^^^^^^^^^
  File "timedeltas.pyx", line 1820, in pandas._libs.tslibs.timedeltas.Timedelta.__new__
  File "timedeltas.pyx", line 653, in pandas._libs.tslibs.timedeltas.parse_timedelta_string
ValueError: unit abbreviation w/o a number
```

The current issue appears to arise as a result of xarray not entirely deprecating loffset (as of version 2023.10.1) so invalid arguments are internally passed to pandas (as of version 2.1.2) which has deprecated loffset

**Comment by kafitzgerald** [comment]
Looks like handling of the time offset arithmetic is different for cftime. 

Probably why `xr.coding.cftime_offsets.to_offset` exists (though this is also private API).

**Comment by kafitzgerald** [comment]
> The current issue appears to arise as a result of xarray not entirely deprecating loffset (as of version 2023.10.1 with https://github.com/pydata/xarray/pull/7444) so invalid arguments are internally passed to pandas (as of version 2.1.2) which has deprecated loffset

I think this is a little more nuanced. 

The xarray fix to postpone the deprecation of the loffset argument should still work for most cases, but [a later PR introduced a line](https://github.com/pydata/xarray/pull/7206/files#diff-71ee1ab3258dca8894ad22fd08ff181e140262a5ac38985d49ee0a6ee6039319R155) that converts the loffset argument to a pandas timedelta (via pd.to_timedelta) and that fails since "MS" is invalid input.  

Unfortunately, this means that the xarray resample/loffset fix no longer works for us and as mentioned above and shown in the CI failures the time offset arithmetic is a little more complicated in this case.

**Review by jukent — APPROVED** [review]

**Review by anissa111 — CHANGES_REQUESTED** [review]
Don't forget the pin in `setup.cfg`!

**Comment by kafitzgerald** [comment]
Having missed some version pins myself recently...  I think this version pin is also in some additional *.yml files and the requirements.txt in addition to the setup.cfg

**Comment by cyschneck** [comment]
The changes to `interpolation.py` are to [fix the issue with the interpolation tests throwing](https://github.com/NCAR/geocat-comp/pull/492#issuecomment-1778087254) 

```
ValueError: new dimensions ('sigma',) must be a superset of existing dimensions ('dim_0',)
```

**Review by kafitzgerald — CHANGES_REQUESTED** [review]
The changes here all look reasonable to me (introducing more private API from xarray is non-ideal, but should be less of a risk with expanded upstream testing and hopefully we do something different in the long run).    

I do think we should better document the change in interpolation.py and the loffset change.  At least including them in the PR description and changing the title and release notes blurb to something more inclusive of the changes here.  

Probably a good idea to open another issue about Xarray private API as well since we'll be introducing more rather than removing it.  

Will be nice to see this get merged though!

File: geocat/comp/interpolation.py [hunk 1]
```diff
@@ -593,19 +593,19 @@ def interp_sigma_to_hybrid(data: xr.DataArray,
         output = data_stacked[:, :len(hyam)].copy()
 
         for idx, (d, s) in enumerate(zip(data_stacked, sigma_stacked)):
-            output[idx, :] = xr.DataArray(
-                _vertical_remap(func_interpolate, s.data, sig_coords.data,
-                                d.data))
+            output[idx, :] = xr.DataArray(_vertical_remap(
+                func_interpolate, s.data, sig_coords.data, d.data),
+                                          dims=[lev_dim])
```
[the remark this answers is not in the document]
Comment (cyschneck) [reply geocat/comp/interpolation.py:598]: Will do!

File: geocat/comp/climatologies.py [hunk 2]
```diff
@@ -302,7 +303,19 @@ def month_to_season(
...
+                  xr.coding.cftimeindex.CFTimeIndex):
+        means[time_coord_name] = means.indexes[
+            time_coord_name] + xr.coding.cftime_offsets.to_offset(freq="MS")
+    elif isinstance(means.indexes[time_coord_name], pd.DatetimeIndex):
+        means[time_coord_name] = means.indexes[
+            time_coord_name] + pd.tseries.frequencies.to_offset(freq="MS")
+    else:
+        raise ValueError(
+            f"unsupported array type - {type(means.indexes[time_coord_name])}. Valid types include: (xr.coding.cftimeindex.CFTimeIndex, pandas.core.indexes.datetimes.DatetimeIndex)"
+        )
```
[the remark this answers is not in the document]
Comment (cyschneck) [reply geocat/comp/climatologies.py:318]: I added a write up in a backlog Issue #508

File: docs/release-notes.rst [hunk 3]
```diff
@@ -23,12 +23,14 @@ Documentation
 Bug Fixes
 ^^^^^^^^^
 * Fix Python version in upstream CI by `Philip Chmielowiec`_ in (:pr:`436`)
+* Unpin xarray in enviroment builds with changes to interpolation.py (specify dims in xr.DataArray) and climatologies.py (replace loffset with to_offset) by `Cora Schneck`_ in (:pr:`381`)
```
Comment (kafitzgerald) [inline docs/release-notes.rst:26]: I think this still needs to get bumped up to a v2023.11.0 (unreleased) section.  

Otherwise, things look good to me!

**Review by kafitzgerald — APPROVED** [review]
Looks good!

**Review by anissa111 — APPROVED** [review]
Sorry, I missed the re-review request, this looks good!

**Pull request merged**